### PR TITLE
Allow <Exec> commands to inherit the console code page

### LINKED ISSUE #4870: MSBuild <Exec>/ToolTask resets the console output code page, prohibiting unicode output
### Steps to reproduce

Repro project (save as `repro.proj`):
```xml
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <Target Name="Print">
        <Exec Command="chcp" />
    </Target>       
</Project>
```

Repro commands in `cmd.exe`:
```
>chcp
Active code page: 850

>msbuild repro.proj
Microsoft (R) Build Engine version 16.4.0-preview-19517-01+31c846a51 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 11/1/2019 8:35:23 AM.
Project "repro.proj" on node 1 (default targets).
Print:
  chcp
  Active code page: 850
Done Building Project "repro.proj" (default targets).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.18

>chcp 65001
Active code page: 65001

>msbuild repro.proj
Microsoft (R) Build Engine version 16.4.0-preview-19517-01+31c846a51 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 11/1/2019 8:35:29 AM.
Project "repro.proj" on node 1 (default targets).
Print:
  chcp
  Active code page: 850
Done Building Project "repro.proj" (default targets).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.23
```

Note: 850 is the Western Europe/US code page, representing most latin-based characters. 65001 is UTF-8, which can represent all Unicode characters.

### Expected  behavior
When run inside MSBuild, `chcp` should output the same code page as its ancestor `cmd.exe`.

Programs should inherit the console code page from their ancestors.

### Actual behavior
When run inside MSBuild, `chcp` outputs 850 rather than 65001 as set on the ancestor `cmd.exe`.

MSBuild's `<Exec>` task executes its commands in such a way that the child processes do not inherit the console code page from MSBuild (which itself inherited it from its parent). The console code page for children processes is rather reset to the OEM code page, which cannot encode Unicode characters in any Windows OS language in its default configuration. This means that it is impossible to use MSBuild to run a program that will output Japanese characters on an English OS without it coming out as a bunch of `?????`.

Specifically, the use of `ProcessStartInfo.CreateNoWindow = true` in `ToolTask` will reset the console code page of the child process to the OEM code page. If this is fixed, the batch file should also be written in the console code page since that is what `cmd.exe` will be using to read it. Pull request coming soon.

### Environment data
`msbuild /version` output: `Microsoft (R) Build Engine version 16.4.0-preview-19517-01+31c846a51 for .NET Framework`
OS info: Windows 10

### PR DESCRIPTION
**Pull request by MrTrillian**
MSBuild's `<Exec>` task executes its commands in such a way that the child processes do not inherit the console code page from MSBuild (which itself inherited it from its parent). The console code page for children processes is rather reset to the OEM code page, which cannot encode Unicode characters in any Windows OS language in its default configuration. This means that it is impossible to use MSBuild's `<Exec>` task to run a program that will output Japanese characters on an English OS without it coming out as a bunch of `?????`.

Specifically, the use of `ProcessStartInfo.CreateNoWindow = true` in `ToolTask` will reset the console code page of the child process to the OEM code page. With this fixed, the batch file must also be written in the console code page since that is what `cmd.exe` will be using to read it.

With this change, a parent process can set the console output code page to 65001 (UTF-8) and have some hope of getting Unicode data back from the programs started by MSBuild.

This change has had minimal testing. I only built it and verified that `chcp` called in `<Exec>` will now output the console code page as inherited from MSBuild's parent process. There is also a risk of regression since some ill-behaved programs always output in the OEM code page, no matter what the console output code page is, and with this change they will have to face the reality of their misdeeds. Regressions would only happen when an ancestor process explicitly changes the console output code page from the default OEM code page to something else.

Fixes #4870

### CODE REVIEW

File: src/Utilities/ToolTask.cs [hunk 1]
```diff
@@ -588,8 +588,20 @@ string responseFileSwitch
...
+            // This is problematic because processes normally write to their standard output using
+            // the console output code page. If that is reset to the default OEM code page,
+            // then it will be impossible to represent all Unicode characters,
+            // since the default configuration for all Windows OS languages is a non-Unicode-capable OEM code page.
+            //
+            // The only way to get Unicode output through the console is if an ancestor process
+            // set the console output code page to UTF-8 and text-producing programs honor it.
             ProcessStartInfo startInfo = new ProcessStartInfo(pathToTool, commandLine);
-            startInfo.CreateNoWindow = true;
+            startInfo.CreateNoWindow = false;
```
Comment (MrTrillian) [inline src/Utilities/ToolTask.cs]: Note: I do not know the ramifications of changing this outside of what I'm trying to fix, if any. The [documentation](https://docs.microsoft.com/en-us/windows/win32/procthread/process-creation-flags) says:

> CREATE_NO_WINDOW 0x08000000
> The process is a console application that is being run without a console window. Therefore, the console handle for the application is not set. This flag is ignored if the application is not a console application, or if it is used with either CREATE_NEW_CONSOLE or DETACHED_PROCESS.

**Comment by MrTrillian** [comment]
Then let us hear from @rainersigwald :)

File: src/Utilities/ToolTask.cs (hunk 1 above)
Comment (MrTrillian) [reply src/Utilities/ToolTask.cs]: If this is problematic, I propose an alternate solution in my comment on #4870

**Comment by rainersigwald** [comment]
This definitely makes me very nervous. Last time we mucked with cmd.exe codepages it caused a year-long bug tail and didn't fix things.

Doesn't avoiding this API cause the executed script to have access to MSBuild's console? That seems like a pretty bad idea: tools could then directly emit output and overwrite the build log.

Have you discussed this change with the Windows console folks?

**Comment by MrTrillian** [comment]
I meant this PR to be a reference to what a fix might look like, but the main point is to make sure the original internal bug gets tracked, and I realize that getting into the implementation is distracting from this goal. I will reassign that internal bug to your path and update the github issue to better reflect it.

**Pull request closed without merging**